doc/random.adoc

```diff
@@ -0,0 +1,193 @@
+
+`num_trials` defaults to `6`. The more trials, the better the approximation.
+
+The return value is a float.
+
+=== `PseudoRandom`
+
+A seedable 16-bit unsigned integer pseudo-random number generator.
+
+"Uses a well-known LCG algorithm introduced by K&R."
+
+Perhaps the lowest-quality random generator of all.
+
+==== `PseudoRandom(seed)`
+
+Constructor: Takes a `seed` and returns a `PseudoRandom` object.
+
+===== `:next([min, max])`
+
+If `min` and `max` are both omitted, they default to `0` and `2^16-1` (`32767`) respectively.
+
+WARNING: Requires `((max - min) == 32767) or ((max-min) <= 6553))` for a proper distribution.
+
+=== `SecureRandom`
+
+System-provided cryptographically secure random:
+An attacker should not be able to predict the generated sequence of random numbers.
+Use this when generating cryptographic keys or tokens.
+
+Not necessarily available in all builds and on all platforms.
+
+==== `SecureRandom()`
+
+Constructor: Returns a SecureRandom object or `nil` if no secure random source is available.
+
+TIP: Use `assert(SecureRandom(), "no secure random available")` to error if no secure random source is available.
+
+==== Methods
+
+===== `:next_bytes([count])`
+
+Only argument is `count`, an optional integer defaulting to `1`
+and limited to `2048` specifying how many bytes are to be returned.
+Returned as a Lua bytestring of length `count`
+
+== Benchmarking
+
+[source,lua]
+---
+collectgarbage"stop" -- we don't want GC heuristics to interfere
+
+local n = 1e8 -- number of runs
+local function bench(name, constructor, invokation)
+	local func = assert(loadstring(([[
+local r = %s
+for _ = 1, %d do %s end
+]]):format(constructor, n, invokation)))
+	local t = minetest.get_us_time()
+	func()
+	print(name, (minetest.get_us_time() - t) / n, "µs/call")
+end
+
+bench("Lua", "nil", "math.random()")
+bench("PCG", "PcgRandom(42)", "r:next()")
+bench("K&R", "PseudoRandom(42)", "r:next()")
+bench("Secure", "assert(SecureRandom())", "r:next_bytes()")
+---
+
+Example output:
+
+[source]
+---
+Lua	0.00385002	µs/call
+PCG	0.05579729	µs/call
+K&R	0.05859349	µs/call
+Secure	0.11211887	µs/call
+---
+
+== Comparison
+
+|===
+| Random Source  | Performance      | Bytes of entropy | Seedability                    | Versatility | Distribution | Security | Portability
+| `math.random`  | very good (1x)   | up to 4          | global seed; seeded by default | very good   | no guarantees, but usually decent enough | not cryptographically secure | varies by platform
+| `PcgRandom`    | okay (~14x)      | up to 4          | per-instance seed | very good   | good, decent guarantees | not cryptographically secure | always the same
+| `PseudoRandom` | okay (~15x)      | 1 to 2           | per-instance seed | outright sucks | okay-ish | not cryptographically secure | always the same
+| `SecureRandom` | still okay (30x) | 1 to 2048        | not seedable | cryptographically secure | varies by platform; may be missing
```

Comment:
too few columns on this row? markdown-style tables are messy with so many columns when they are not properly formatted, [asciidoc supports a more ergonomic table syntax](https://github.com/minetest/minetest_docs/pull/17/files#diff-ee860ca877adc533b0369fb3fa42a24943010a41f40d49c84aa760da1ee0f49fR11) that's a good idea to use in these cases